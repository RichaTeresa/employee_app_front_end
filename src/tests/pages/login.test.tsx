
import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import userEvent from "@testing-library/user-event";
import Login from "../../pages/login/login";
const mockNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});// Mock the store
const mockStore = configureStore({
  reducer: {
    employee: (state = { employees: [] }) => state,
    department: (state = { departments: [] }) => state,
    employeeApi: (state = {}) => state,
  },
});// Mock the login mutation
const mockLoginMutation = vi.fn();
vi.mock("../../api-service/auth/login.api", () => ({
  useLoginMutation: () => [
    () => ({
      unwrap: () => mockLoginMutation(),
    }),
    { isLoading: false },
  ],
}));describe("Login Page", () => {
  const renderLogin = () => {
    return render(
      <Provider store={mockStore}>
        <Login />
      </Provider>
    );
  };  beforeEach(() => {
    vi.clearAllMocks();
  });  it("should match snapshot", () => {
    const { container } = renderLogin();
    expect(container).toMatchSnapshot();
  });

  it("should render login form with all required elements", () => {
    renderLogin();

    // Check if username input is present
    const usernameInput = screen.getByPlaceholderText("username");
    expect(usernameInput).toBeInTheDocument();

    // Check if password input is present
    const passwordInput = screen.getByPlaceholderText("password");
    expect(passwordInput).toBeInTheDocument();

    // Check if submit button is present
    const submitButton = screen.getByRole("button", { name: /logging in/i });
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toHaveAttribute("type", "submit");
  });

  it("should show error when username is too long", async () => {
    renderLogin();
    const usernameInput = screen.getByPlaceholderText("username");    // Type a username longer than 30 characters
    await userEvent.type(usernameInput, "a".repeat(31));    // Check if error message is displayed
    expect(
      screen.getByText("Username must not be more than 10 characters!")
    ).toBeInTheDocument();
  });

  

  it("should handle successful login", async () => {
    mockLoginMutation.mockResolvedValueOnce({ token: "fake-token" });

    renderLogin();

    // Fill in the form
    const usernameInput = screen.getByPlaceholderText("username");
    const passwordInput = screen.getByPlaceholderText("password");
    const submitButton = screen.getByRole("button", { name: /logging in/i });

    await userEvent.type(usernameInput, "test@example.com");
    await userEvent.type(passwordInput, "password123");

    // Submit the form
    await userEvent.click(submitButton);

    // Wait for the login mutation to resolve and navigation to occur
    await waitFor(() => {
      expect(mockLoginMutation).toHaveBeenCalled();
      expect(mockNavigate).toHaveBeenCalledWith("/employees");
    });
  });
});