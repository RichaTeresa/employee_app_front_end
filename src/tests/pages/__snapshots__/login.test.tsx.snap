// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Page > should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="leftside"
    >
      <img
        alt="kv-login"
        class="circle"
        src="/src/assets/kv-login.jpeg"
      />
    </div>
    <div
      class="rightside"
    >
      <p>
        X coordinate:
        0
         Y coordinate :
        0
      </p>
      <div
        class="login-box"
      >
        <img
          alt="kv-logo"
          class="logo"
          src="/src/assets/kv-logo.png"
        />
        <div
          class="form-input-2"
        >
          <label
            for="username"
          />
          <div
            class="add-adornment"
          >
            <input
              id="username"
              placeholder="username"
              required=""
              type="text"
              value=""
            />
            <button
              disabled=""
            >
              clear
            </button>
          </div>
        </div>
        <div
          class="form-input-2"
        >
          <label
            for="password"
          />
          <div
            class="add-adornment"
          >
            <input
              id="password"
              placeholder="password"
              required=""
              type="password"
              value=""
            />
            <button
              disabled=""
            >
              clear
            </button>
          </div>
        </div>
        <div
          class="form-input-2"
        >
          <label
            for="showPassword"
          >
            show password
          </label>
          <div
            class="add-adornment"
          >
            <input
              id="showPassword"
              required=""
              type="checkbox"
            />
          </div>
        </div>
        <p
          style="color: red;"
        />
        <button
          class="login-button"
          id="login"
          type="submit"
        >
          logging in
        </button>
      </div>
    </div>
  </div>
</div>
`;
